exercises/dolica_akelloegwel/ex02_oo_basics/src/Shape.h

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <string>
+#include <iostream>
+#include <math.h>
+#include <vector>
+#include <algorithm>
+
+class Shape
+{
+protected:
+	double m_area;
+	double m_perimeter;
+	std::string m_type;
+	int m_numSides;
+	std::string m_printMessage;
+	void SetPerimeter(double perimeter);
+	void SetArea(double area);
+	void SetSides(int sides);
+	std::string CreateBaseMessage();
+	void SetMessage(std::string msg);
+
+public:
+	Shape(std::string t, int n);
+	~Shape();
+	double GetPerimeter() const;
+	double GetArea() const;
+	std::string GetType() const;
+	int GetSides() const;
+	std::string GetMessage() const;
+	friend std::ostream &operator<<(std::ostream & os, const Shape &shape);

```

Comment:
Interesting solution here. You survive the slicing because you perform the calculations in the constructor. Here are a few problems:

1. Your solution would not work if say your shape was mutable. i.e. change radius of circle shape then get area.
1. You can construct a "Shape" stand alone because it has a public constructor and is not pure-virtual. What does an object of type "Shape" mean? Can I have a Shape that isn't really any Shape?

Also your Shape is type is not just contractural "fulfil these methods I don't care how", but implicitly implies how it can be done.